components/shipcloud/app/shipcloud.app.ts

```diff
@@ -0,0 +1,161 @@
+import { defineApp } from "@pipedream/types";
+import { axios } from "@pipedream/platform";
+import {
+  CreateHookParams,
+  CreateShipmentParams,
+  CreateShipmentQuoteParams,
+  DeleteHookParams,
+  GetShipmentParams, HttpRequestParams,
+} from "../common/requestParams";
+import {
+  Address, Shipment, ShipmentQuote, Webhook,
+} from "../common/responseSchemas";
+import {
+  CARRIER_OPTIONS, SERVICE_OPTIONS,
+} from "../common/constants";
+import { OptionsObject } from "../common/types";
+
+export default defineApp({
+  type: "app",
+  app: "shipcloud",
+  methods: {
+    _baseUrl(): string {
+      return "https://api.shipcloud.io/v1";
+    },
+    async _httpRequest({
+      $ = this,
+      endpoint,
+      ...args
+    }: HttpRequestParams): Promise<object> {
```

Comment:
Humm, since we are using OO, I am wondering if we could use Generics instead of return an `object`
https://www.typescriptlang.org/docs/handbook/2/generics.html

Reply:
Thanks for the suggestion! Would this be different than defining the return type on each of the request methods (such as `listShipments` and `listAddresses`)?

Currently, the only method that returns `_httpRequest` without defining an explicit return type is `deleteHook` - because that returns `204 No Content`. How do you think this could be improved with generics? @vellames

Reply:
It would not have a super improvement, just adding more meaning to the returns.
Instead of returning `Promise<object>` you could return `Promise<T>` and use generics.
So the signature of `httpRequest` would be something like `httpRequest<T>(...) : Promise<T>`.
Using this you can do the following in the methods calling `httpRequest`:

```
async _httpRequest<T>({
      $ = this,
      endpoint,
      ...args
    }: HttpRequestParams): Promise<T> {
      return axios($, {
        url: this._baseUrl() + endpoint,
        auth: {
          username: this.$auth.api_key,
          password: "",
        },
        ...args,
      });
    },
...
async createShipment(params: CreateShipmentParams): Promise<Shipment> {
      return this._httpRequest<Shipment>({
        method: "POST",
        endpoint: "/shipments",
        ...params,
      });
    },
```

Reply:
I see, passing a type argument indeed seems much better than simply defining the method return types!

Unfortunately, this can't be implemented with the current TypeScript structure for PD app components... in the example you suggested, `this._httpRequest(...)` is not recognized and is treated as an untyped function call, so I can't pass a type argument.

![image](https://user-images.githubusercontent.com/48412907/185227768-383ecd54-9960-4716-b2d5-052f0172fbc7.png)
![image](https://user-images.githubusercontent.com/48412907/185227853-12d25413-4ca4-4d42-8597-4368da18ab54.png)

Reply:
Humm. I see, strange, okay :)